### PR TITLE
MRG : add support for vol source spaces in spatial_src_connectivity

### PR DESCRIPTION
**Pull request by agramfort**
I guess I should add support in spatio_temporal_src_connectivity too.

maybe tomorrow...

my debug script : https://gist.github.com/agramfort/df1c55fd60682680eb8751207d24f6b5

### CODE REVIEW

**Comment by agramfort** [comment]
should be ready provided CIs are happy

current code does not work with mixed source spaces.

@larsoner we should be using https://docs.scipy.org/doc/scipy-0.16.0/reference/generated/scipy.sparse.block_diag.html to assemble the connectivity from all source spaces. I would compute connectivity for each source space (each hemi) and make a block diagonal matrix from it. WDYT?

**Comment by larsoner** [comment]
> I would compute connectivity for each source space (each hemi) and make a block diagonal matrix from it. WDYT?

Yep, sounds good to me

File: mne/source_estimate.py [hunk 1]
```diff
@@ -2639,14 +2639,64 @@ def morph_data_precomputed(subject_from, subject_to, stc_from, vertices_to,
...
+def _spatio_temporal_src_connectivity_vol(src, n_times):
+    assert len(src) == 1  # not a mixed source space
+    from sklearn.feature_extraction import grid_to_graph
+    vertices = np.where(src[0]['inuse'])[0]
+    n_vertices = len(vertices)
+    data = (1 + np.arange(n_vertices))[:, np.newaxis]
+    stc_tmp = VolSourceEstimate(data, vertices, tmin=0., tstep=1.)
+    img = stc_tmp.as_volume(src, mri_resolution=False)
+    img_data = img.get_data()[:, :, :, 0]
+    img_data = img_data.T
```
Comment (larsoner) [inline mne/source_estimate.py]: I suspect that these lines can be replaced by direct logic on the source space, rather than doing a temp STC, volume export, then data grab. Have you looked?

File: mne/source_estimate.py [hunk 2]
```diff
@@ -2639,14 +2639,64 @@ def morph_data_precomputed(subject_from, subject_to, stc_from, vertices_to,
...
+    edges = grid_to_graph(*img_data.shape, mask=(img_data != 0))
+    connectivity = _get_connectivity_from_edges(edges, n_times)
+    return connectivity
+
+
+def _spatio_temporal_src_connectivity_ico(src, n_times):
+    if src[0]['use_tris'] is None:
+        raise RuntimeError("The source space does not appear to be an ico "
+                           "surface. Connectivity cannot be extracted from"
+                           " non-ico source spaces.")
```
Comment (larsoner) [inline mne/source_estimate.py]: This implies `oct` also cannot be used, but it probably can. (I know this is from old code but maybe we should fix it here?)

File: mne/source_estimate.py [hunk 3]
```diff
@@ -2666,41 +2716,18 @@ def spatio_temporal_src_connectivity(src, n_times, dist=None, verbose=None):
...
-            # return to original format
-            connectivity = connectivity.tocoo()
-
-        return connectivity
-    else:  # use distances computed and saved in the source space file
-        return spatio_temporal_dist_connectivity(src, n_times, dist)
+    if src[0]['type'] == 'vol':
+        if dist is not None:
+            raise ValueError('dist has to be None for a volume '
+                             'source space. Got %s.' % dist)
```
Comment (larsoner) [inline mne/source_estimate.py]: Should have doc update saying `dist` must be None if there are any vol source spaces

Also probably update the doc to say volume and surface source spaces are supported, but not discrete.

File: mne/source_estimate.py (hunk 1 above)
Comment (larsoner) [reply mne/source_estimate.py]: (This would also increase the number of volume source spaces that can be used to those without MRI, like volume source space within a generic sphere, because `as_volume` requires MRI vol source space)

File: mne/source_estimate.py (hunk 3 above)
Comment (larsoner) [reply mne/source_estimate.py]: Just noticed you already have the latter change. But the "`dist` must be None" change should still go in there.

And if you do block diag for mixed it'll need update, too

File: mne/source_estimate.py (hunk 1 above)
Comment (agramfort) [reply mne/source_estimate.py]: no I have not looked. I will just put an XXX note for later thing.

File: mne/source_estimate.py (hunk 2 above)
Comment (agramfort) [reply mne/source_estimate.py]: I've put an XXX TODO :)

File: mne/tests/test_source_estimate.py [hunk 4]
```diff
@@ -881,4 +886,24 @@ def test_vec_stc():
     assert_array_equal(normal.data[:, 0], [1, 2, 0, np.sqrt(3)])
 
 
+@requires_sklearn
+@requires_nibabel
```
Comment (larsoner) [inline mne/tests/test_source_estimate.py]: This must be called as a function, it's why Travis is angry

**Comment by agramfort** [comment]
green finally...

**Comment by larsoner** [comment]
[Pending question](https://github.com/mne-tools/mne-python/pull/4544#pullrequestreview-63212347):
> Do you want me to look on Monday about avoiding the nibabel input before merge and push directly to your branch?

**Comment by agramfort** [comment]
Please do if you have some spare cycles

File: mne/source_estimate.py [hunk 5]
```diff
@@ -2639,14 +2639,70 @@ def morph_data_precomputed(subject_from, subject_to, stc_from, vertices_to,
...
+    data = (1 + np.arange(n_vertices))[:, np.newaxis]
+    # XXX these lines can certainly be replaced by direct logic on the source
+    # space, rather than doing a temp STC, volume export, then data grab. Have
+    # you looked? This would also increase the number of volume source spaces
+    # that can be used to those without MRI, like volume source space within
+    # a generic sphere, because as_volume requires MRI vol source space)
+    stc_tmp = VolSourceEstimate(data, vertices, tmin=0., tstep=1.)
+    img = stc_tmp.as_volume(src, mri_resolution=False)
+    img_data = img.get_data()[:, :, :, 0]
+    img_data = img_data.T
```
Comment (larsoner) [inline mne/source_estimate.py]: @agramfort why is this `.T` here? It makes it so that the `img_data.shape != src[0]['shape']` (which it otherwise does without the `.T` there)

Comment (larsoner) [reply mne/source_estimate.py]: ahh never mind, I see it now (necessary to get `np.array_equal(img_data[mask], data[:, 0])`)

**Comment by larsoner** [comment]
Okay @agramfort take a look at the simplification and test see if they make sense to you

File: mne/tests/test_source_estimate.py [hunk 6]
```diff
@@ -881,4 +886,45 @@ def test_vec_stc():
...
+    assert_true(np.all(connectivity.data == 1))
+    assert_true(isinstance(connectivity, sparse.coo_matrix))
+
+    connectivity2 = spatio_temporal_src_connectivity(vol, n_times=2)
+    assert_equal(connectivity2.shape, (2 * n_vertices, 2 * n_vertices))
+    assert_true(np.all(connectivity2.data == 1))
+
+
+@requires_sklearn
+@requires_nibabel()
```
Comment (agramfort) [inline mne/tests/test_source_estimate.py]: @larsoner remove nibabel?

Comment (larsoner) [reply mne/tests/test_source_estimate.py]: No the `vol_mask` test uses `nibabel` because it uses `as_volume`. I did remove it for the test above this one, which no longer needs it.

**Comment by agramfort** [comment]
did you check that it gives the same matrix as before?

**Comment by larsoner** [comment]
That is what the new `test_vol_mask` function does, it uses your code and checks the code I wrote against it

**Comment by agramfort** [comment]
great. Then merge when Cis are green

thx heaps
